### ISSUE TITLE
[TF1.13] gate order in tf.keras.layers.cudnnlstm

### ISSUE THREAD

**Issue by edchengg** [tensorflow/tensorflow#29308]
There is a kernel [input_dim, hid_dim * 4] matrix and recurrent kernel [hid_dim, hid_dim * 4] matrix in cudnnlstm.

**My question is what is the order of gate input, forget, new, output in the kernel and recurrent kernel?
[input, forget, new, output]**

I would really appreciate it since I am transfering weights from BasicLSTMCell kernel matrix.
I would also like to confirm the gate order in BasicLSTMCell kernel [input_dim + hid_dim, hid_dim * 4] is 
[input, new, forget, output]

Thank you!

**Comment by gadagashwini** [comment]
@edchengg Could you provide more details about the issue and context? Thanks!

**Comment by edchengg** [comment]
@gadagashwini 

I am referring to 
https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/keras/layers/cudnn_recurrent.py

L473:

`params = recurrent._canonical_to_params(    # pylint: disable=protected-access
        weights=[
            self.kernel[:, :self.units],
            self.kernel[:, self.units:self.units * 2],
            self.kernel[:, self.units * 2:self.units * 3],
            self.kernel[:, self.units * 3:],
            self.recurrent_kernel[:, :self.units],
            self.recurrent_kernel[:, self.units:self.units * 2],
            self.recurrent_kernel[:, self.units * 2:self.units * 3],
            self.recurrent_kernel[:, self.units * 3:],
        ],
        biases=[
            self.bias[:self.units],
            self.bias[self.units:self.units * 2],
            self.bias[self.units * 2:self.units * 3],
            self.bias[self.units * 3:self.units * 4],
            self.bias[self.units * 4:self.units * 5],
            self.bias[self.units * 5:self.units * 6],
            self.bias[self.units * 6:self.units * 7],
            self.bias[self.units * 7:],
        ],
        shape=self._vector_shape)`

It is not clear which gate(input, forget, new, output) params is self.kernel[:, :] referring to in the cudnnLSTM.

Thank you.

**Comment by qlzh727** [comment]
There are 3 fused weights in keras LSTM layer, kernel, recurrent_kernel, and bias. Each of the fused weights are in IFCO order (input, forget, carry, output). 

BasicLSTMCell is existing TF code have two fused weights: kernel and bias. The kernel here is the combination of kernel and recurrent kernel as keras layer, which is concat on dim 0. More importantly, the order of the fused kernel is ICFO (input, carry, forget, output).

If you want to converting your BasicLSTMCell weights to keras format, you need to first split the kernel into 2 parts, and also switch the carry and forget gate.

**Comment by edchengg** [comment]
I have solved this issue by myself.
The gate order in basiclstmcell is [input, carry, forget, output]
the gate order in tf.keras.layers.cudnnlstm is [input, forget, carry, output]

We need to further convert the weight matrix for each gate when transferring from basiclstmcell to tf.keras.layers.cudnnlstmcell.

With square matrix: we can use transpose.
With matrix with size [row, col] (for example, input kernel)
we need to do the following:
[[0, 1, 2, 3],                [[0, 4, 1, 5],
  [4, 5, 6, 7]] .     -->  [2, 6, 3, 7]]

**Comment by zhuzilin** [comment]
@edchengg Thank you so much!
